crates/astria-conductor/src/executor/queue.rs

```diff
@@ -0,0 +1,683 @@
+use std::collections::{
+    BTreeMap,
+    HashMap,
+};
+
+use color_eyre::eyre::Result;
+use tendermint::{
+    block::Height,
+    hash::Hash,
+};
+use tracing::info;
+
+use crate::types::SequencerBlockSubset;
+
+/// A queue for the SequencerBlockSubset type that holds blocks that are
+/// pending or not yet ready for execution.
+///
+/// This Queue handles all the fork choice logic for incoming Sequencer blocks
+/// from gossip. It is responsible for determining which blocks are safe to
+/// pass on to execution, and holds on to the other data until it is safe to
+/// execute, or deletes it if it is no longer needed or becomes stale or invalid.
+#[derive(Debug, Clone)]
+pub(super) struct Queue {
+    head_height: Height,
+    most_recent_soft_hash: Hash,
+
+    // The collection of all pending blocks. the blocks in this map at Height ==
+    // Queue.head_height are the head blocks
+    pending_blocks: HashMap<Height, HashMap<Hash, SequencerBlockSubset>>,
+    // All blocks in order by height that can be considered safe because they
+    // have a child block
+    soft_blocks: BTreeMap<Height, SequencerBlockSubset>,
+}
+
+impl Queue {
+    pub(super) fn new() -> Self {
+        Self {
+            head_height: Height::default(),
+            most_recent_soft_hash: Hash::default(),
+            pending_blocks: HashMap::new(),
+            soft_blocks: BTreeMap::new(),
+        }
+    }
+
+    /// Inserts a new block into the ExecutorQueue.
+    ///
+    /// This is the only way to add data to the queue. When inserting blocks,
+    /// the internal state of the queue will also be updated to properly order
+    /// and arrange all blocks in the queue, based on the Tendermint/CometBFT fork
+    /// choice rules.
+    pub(super) fn insert(&mut self, block: SequencerBlockSubset) -> Result<Option<Hash>> {
+        // if the block is already in the queue, return its hash
+        if self.is_block_present(&block) {
+            info!(
+                block.height = %block.height(),
+                block.hash = %block.block_hash(),
+                "block is already present in the queue"
+            );
+            return Ok(Some(block.block_hash()));
+        }
+
+        // if the block is stale, ignore it
+        if block.header().height < self.head_height {
```

Comment:
the naming here is a bit confusing - only blocks higher than the `head_height` are added to the queue? can you rename `head_height` to `base_height` or `safe_height` something? also from what i understand, all blocks in the queue should be a descendant of this "root" block (the highest safe block), is there a check for that?